src/luaotfload-harf-plug.lua

```diff
@@ -0,0 +1,1000 @@
+    local fontdata = font.getfont(fontid)
+    local options = fontdata.specification.features.raw
+
+    local glyphs
+    head, glyphs = shape(head, current, run)
+    return tonodes(head, current, run, glyphs)
+  else
+    return head, run.after
+  end
+end
+
+function process(head, font, direction)
+  local newhead, current = head, head
+  local runs = itemize(head, font, direction)
+
+  for _, run in next, runs do
+    newhead, current = shape_run(newhead, current, run)
+  end
+
+  return newhead or head
+end
+
+local function pdfdirect(data)
+  local n = newnode(whatsit_t, pdfliteral_t)
+  setfield(n, "mode", 2) -- direct
+  setdata(n, data)
+  return n
+end
+
+local function post_process(head)
+  for n in traverse(head) do
+    local props = getproperty(n)
+    local harfprops = props and props.harf
+
+    local startactual, endactual
+    if harfprops then
+      startactual = harfprops[startactual_p]
+      endactual = harfprops[endactual_p]
+    end
+
+    if startactual then
+      local actualtext = "/Span<</ActualText<FEFF"..startactual..">>>BDC"
+      head = insertbefore(head, n, pdfdirect(actualtext))
+    end
+
+    if endactual then
+      head = insertafter(head, n, pdfdirect("EMC"))
+    end
+
+    local replace = getfield(n, "replace")
+    if replace then
+      setfield(n, "replace", post_process(replace))
+    end
+
+    local subhead = getfield(n, "head")
+    if subhead then
+      setfield(n, "head", post_process(subhead))
+    end
+  end
+  return head
+end
+
+local function post_process_nodes(head, groupcode)
+  return tonode(post_process(todirect(head)))
+end
+
+local function run_cleanup()
+  -- Remove temporary PNG files that we created, if any.
+  -- FIXME: It would be nice if we wouldn't need this
+  for _, path in next, pngcachefiles do
+    osremove(path)
+  end
+end
+
+local function set_tounicode()
+  for fontid, fontdata in font.each() do
+    local hbdata = fontdata.hb
+    if hbdata and fontid == pdf.getfontname(fontid) then
+      local characters = fontdata.characters
+      local newcharacters = {}
+      local hbshared = hbdata.shared
+      local glyphs = hbshared.glyphs
+      local nominals = hbshared.nominals
+      local gid_offset = hbshared.gid_offset
+      for gid = 0, #glyphs do
+        local glyph = glyphs[gid]
+        if glyph.used then
+          local character = characters[gid + gid_offset]
+          newcharacters[gid + gid_offset] = character
+          local unicode = nominals[gid]
+          if unicode then
+            newcharacters[unicode] = character
+          end
+          character.tounicode = glyph.tounicode or "FFFD"
+          character.used = true
+        end
+      end
+      font.addcharacters(fontid, { characters = newcharacters })
+    end
+  end
+end
+
+-- FIXME: Move this into generic parts of luaotfload
+local function get_glyph_info(n)
+  local props = properties[todirect(n)]
+  if not props then return end
+  return props and props.glyph_info or nil
```

Comment:
Since LuaTeX slightly changed the semantics of the callback, and now `nil` will cause, it to print the node `character`, this code needs to be adapted and return `nil` only when it is not a HarfBuzz processed font, otherwise return empty sting (this is to handle characters that map to multiple glyph, the first glyph will print the character and the rest should print nothing).

Reply:
That's handled earlier: The remaining characters have `props.glyph_info` set to the empty string, so for them nothing is printed. The function still has to change to print something sensible for other fonts though.